File: test/init.test.ts
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { afterEach, describe, expect, it } from "vitest";
import { git } from "../src/git.js";
import { initialize } from "../src/init.js";
import { fileExists, readJson, readYaml } from "../src/json.js";
import type { ShipConfig, ShipManifest } from "../src/types.js";

const temporaryDirectories: string[] = [];

afterEach(async () => {
  await Promise.all(
    temporaryDirectories
      .splice(0)
      .map((directory) => fs.rm(directory, { recursive: true, force: true })),
  );
});

describe("initializer", () => {
  it("creates tracked state and the complete workflow", async () => {
    const root = await fs.mkdtemp(path.join(os.tmpdir(), "smf-init-"));
    temporaryDirectories.push(root);
    await fs.mkdir(path.join(root, "ios"));
    await fs.writeFile(
      path.join(root, "pubspec.yaml"),
      "name: example\nversion: 3.2.1+42\n",
    );
    await git(root, ["init", "-b", "main"]);
    await git(root, ["config", "user.name", "Test"]);
    await git(root, ["config", "user.email", "test@example.com"]);
    await git(root, ["add", "."]);
    await git(root, ["commit", "-m", "chore: bootstrap"]);
    const baselineSha = await git(root, ["rev-parse", "HEAD"]);

    await initialize({ root, bundleId: "dev.example.app" });

    const manifest = await readJson<ShipManifest>(
      path.join(root, ".ship-my-flutter", "manifest.json"),
    );
    expect(manifest.platforms.ios).toMatchObject({
      version: "3.2.1",
      baselineSha,
      pendingRelease: false,
    });
    const configPath = path.join(root, ".ship-my-flutter", "config.yaml");
    const config = await readYaml<ShipConfig>(configPath);
    expect(config.platforms.ios.appStore.mode).toBe("upload-only");
    expect(await fs.readFile(configPath, "utf8")).toContain(
      "# yaml-language-server: $schema=https://raw.githubusercontent.com/Ventairy/ship-my-flutter/main/schemas/config.schema.json",
    );
    expect(
      await fileExists(
        path.join(root, ".ship-my-flutter", "candidates", ".gitkeep"),
      ),
    ).toBe(true);
    const workflow = await fs.readFile(
      path.join(root, ".github", "workflows", "ship-my-flutter.yml"),
      "utf8",
    );
    expect(workflow).toContain("Ventairy/ship-my-flutter-action@v1");
    expect(workflow).toContain("runs-on: macos-26");
    expect(workflow.match(/persist-credentials: false/gu)).toHaveLength(3);
  });
});

File: test/init_test.dart
import 'dart:io';

import 'package:path/path.dart' as p;
import 'package:ship_my_flutter/ship_my_flutter.dart';
import 'package:test/test.dart';

void main() {
  test('initializer creates tracked state and the complete workflow', () async {
    final root = await Directory.systemTemp.createTemp('smf-init-');
    addTearDown(() => root.delete(recursive: true));
    await Directory(p.join(root.path, 'ios')).create();
    await File(
      p.join(root.path, 'pubspec.yaml'),
    ).writeAsString('name: example\nversion: 3.2.1+42\n');
    await git(root.path, const <String>['init', '-b', 'main']);
    await git(root.path, const <String>['config', 'user.name', 'Test']);
    await git(root.path, const <String>[
      'config',
      'user.email',
      'test@example.com',
    ]);
    await git(root.path, const <String>['add', '.']);
    await git(root.path, const <String>['commit', '-m', 'chore: bootstrap']);
    final baselineSha = await currentSha(root.path);

    await initialize(InitOptions(root: root.path, bundleId: 'dev.example.app'));

    final manifest = await loadManifest(root.path);
    expect(manifest.ios.version, '3.2.1');
    expect(manifest.ios.baselineSha, baselineSha);
    expect(manifest.ios.pendingRelease, isFalse);
    final config = await loadConfig(root.path);
    expect(config.schemaVersion, 4);
    expect(config.appPath, '.');
    expect(config.ios.buildCommand, isNull);
    expect(config.ios.appStore.mode, ReleaseMode.upload);
    final configText = await File(
      resolveShipPaths(root.path).config,
    ).readAsString();
    expect(
      configText,
      contains(
        '# yaml-language-server: \$schema='
        'https://raw.githubusercontent.com/Ventairy/ship-my-flutter/main/'
        'schemas/config.schema.json',
      ),
    );
    expect(configText, isNot(contains('build_command:')));
    expect(configText, isNot(contains('ipa_output_path:')));
    expect(config.ios.ipaOutputPath, 'build/ios/ipa');
    expect(
      await File(
        p.join(root.path, '.ship-my-flutter', 'candidates', '.gitkeep'),
      ).exists(),
      isTrue,
    );
    final workflow = await File(
      p.join(root.path, '.github', 'workflows', 'ship-my-flutter.yml'),
    ).readAsString();
    expect(workflow, contains('Ventairy/ship-my-flutter-action@v1'));
    expect(workflow, contains('subosito/flutter-action@'));
    expect(workflow, contains("hashFiles('.fvmrc', '.fvm/fvm_config.json')"));
    expect(workflow, contains('dart pub global activate fvm 4.1.2'));
    expect(workflow, contains('runs-on: macos-26'));
    expect(
      RegExp('persist-credentials: false').allMatches(workflow),
      hasLength(3),
    );
  });
}
